File: Go/049. Group Anagrams.go
package main

import	"strconv"

func groupAnagrams(strs []string) [][]string {
	res := make([][]string, 0)
	m := make(map[string][]string)
	for _, str := range strs {
		tmp := helper(str)
		if _, ok := m[tmp]; !ok {
			m[tmp] = make([]string, 0, 1)
		}
		m[tmp] = append(m[tmp], str)
	}
	for _, v := range m {
		res = append(res, v)
	}
	return res
}

func helper(str string) string {
	res := ""
	tmp := make([]int, 26, 26)
	for i := 0; i < 26; i++ {
		tmp[i] = 0
	}
	for i := 0; i < len(str); i++ {
		tmp[str[i] - 'a']++
	}
	for i := 0; i < 26; i++ {
		res += strconv.Itoa(tmp[i])
	}
	return res
}

File: C++/049. Group Anagrams.cpp
#include <iostream>
#include <vector>
#include <map>

class Solution {
public:
    std::vector<std::vector<std::string>> groupAnagrams(std::vector<std::string>& strs) {
        std::vector<std::vector<std::string>> res;
        std::map<std::string, std::vector<std::string>> M;
        for (auto str : strs) {
            std::string nums(26, '0');
            for (int i = 0; i < str.size(); i++) {
                nums[int(str[i] - 'a')]++;
            }
            M[nums].emplace_back(str);
        }
        for (auto i : M) {
            res.emplace_back(i.second);
        }
        return res;
    }
};